features/cellular/framework/device/CellularStateMachine.cpp

```diff
@@ -541,7 +541,7 @@ nsapi_error_t CellularStateMachine::run_to_state(CellularStateMachine::CellularS
 
 void CellularStateMachine::pre_event(CellularState state)
 {
-    if (_target_state < state) {
+    if (_state < state) {
```

Comment:
This change is wrong. _state is current state what is running, _target_state is the state we wan't to reach.
We should update only the target state (state where to stop running). If connect-disconnect-connect fails it's because CellularDevice havent't call _state_machine->reset(); in cellular_callback. I believe this is the issue and should be investigated. I can check #9760.

Reply:
> I believe this is the issue and should be investigated. I can check #9760.

Please do.

Note that I've added
```
_device->cellular_callback(NSAPI_EVENT_CONNECTION_STATUS_CHANGE, NSAPI_STATUS_DISCONNECTED);
```

just before [this](https://github.com/ARMmbed/mbed-os/blob/mbed-os-5.11.4/features/cellular/framework/AT/AT_CellularContext.cpp#L691) line in `AT_CellularContext::disconnect`, which should call [this](https://github.com/ARMmbed/mbed-os/blob/master/features/cellular/framework/device/CellularDevice.cpp#L190), I think.

Reply:
@jarvte I don't know if you rather want to continue the discussion here or in the issue, I'll just post it here but if you rather want it in issue let me know!

Based on your previous comment, I did some more testing. The error I was originally getting was (see [log 141](https://github.com/ARMmbed/mbed-os/files/2899497/0141.log), line 390)

```
pre_event: wanted state is already / will be achieved, return without launching new event
```

Now, when I changed `_target_state` to `_state` that disappeared and I got a connection the 2nd time as well. Your previous comment got me thinking and I looked further in the disconnect.

- `_state_machine->reset` is called from `CellularDevice::cellular_callback` on a disconnect. It is actually called multiple times (see [log 140](https://github.com/ARMmbed/mbed-os/files/2899494/0140.log)) so I don't know it that is the expected behaviour.
- `CellularStateMachine::reset` doesn't reset the `_target_state`.
  When I add `_target_state = STATE_INIT;` here, I get a connection the 2nd time. Is the `_target_state` supposed to be reset too?

Reply:
@jarvte See this change in the new commit I added (c442b66), fixes #9760 for me. Is this change ok?

Reply:
I'm not sure whether adding `_device->cellular_callback(NSAPI_EVENT_CONNECTION_STATUS_CHANGE, NSAPI_STATUS_DISCONNECTED);` is still needed in `AT_CellularContext::disconnect` since I'm on release 5.11.4 and something in the disconnect has changed in master. (It is needed on 5.11.4.)

Reply:
@vidavidorra  _target_state = STATE_INIT; is already done in CellularStateMachine::reset() in line before calling enter_to_state(STATE_INIT); 
I have been busy with other tasks, will look in this today. Sorry for the delay.

Reply:
@jarvte I believe we're looking at the same code but at different versions. I'm running mbed-os-5.11.4 and am generally looking at this code at that version. I believe you're looking at the code at master?



Based on your previous comment, I've looked into the code at code vs 5.11.4 and found that `_target_state = STATE_INIT;` was added after 5.11.4.

- Reset function in master ([link](https://github.com/ARMmbed/mbed-os/blob/d96c9e88009f718c2c9c78160753a5675dd9e4ae/features/cellular/framework/device/CellularStateMachine.cpp#L81-L91))
- Reset function in 5.11.4 ([link](https://github.com/ARMmbed/mbed-os/blob/mbed-os-5.11.4/features/cellular/framework/device/CellularStateMachine.cpp#L81-L90))

Based on that being added in master tells me this is probably already fixed. I've now also added [this](https://github.com/ARMmbed/mbed-os/issues/9760#issuecomment-465608892) bit to add the cellular callback to a disconnect call but I don't know whether that is still needed in the code that is now on master. Can you check?

- Looks like this was added in 725e14d15f61aa6e9478e4169bd1912bbaef3f07 (PR #9568), so should be in 5.12.0 right? Any idea when 5.12.0 will be released?
- Is it safe to use the code in master for my production device? I'll definitely use it to test whether the code at master fixes the issues I had, but am curious whether the changes in master are already extensively tested and safe to use.

Reply:
I will check. I was looking this PR "Files changed", there you can see that _target_state = STATE_INIT; is added twice. You need to expand the file.

Reply:
> I was looking this PR "Files changed", there you can see that _target_state = STATE_INIT; is added twice. You need to expand the file

Oh, I see. I think the problem is that I'm running this, and making the changes, in mbed-os-5.11.4 but for the PR add the changes in my fork from master. My bad. 
I'll run my code with what is currently on master and post back the results of that. That should also tell us whether any of the changes in this PR are actually necessary.

Reply:
reset() still has ` _target_state = STATE_INIT;` twice.